### ISSUE TITLE
AdMob Rewarded Video early exit not supported consistently

### ISSUE THREAD

**Issue by joeykrim** [mopub/mopub-android-mediation#104]
## Early exit: If a user exits an AdMob Rewarded Video before it's finished:
Screenshots of the early exit opportunity at the end of this issue
### 1) Unable to reload as unable to complete request to MoPubRewardedVideos.loadRewardedVideo()
Calls to MoPubRewardedVideos.loadRewardedVideo() inside of the returned onRewardedVideoClosed() throw this debug message, ["Did not queue rewarded ad request for ad " + "unit %s. This ad unit already finished loading and is ready to show."](https://github.com/mopub/mopub-android-sdk/blob/6f8c457cf599fe2ae038abd6008b929ea5e48732/mopub-sdk/mopub-sdk-rewardedvideo/src/main/java/com/mopub/mobileads/MoPubRewardedVideoManager.java#L319)
### 2) Error thrown from MoPubRewardedVideos.showRewardedVideo()
The next subsequent call, after the early exit, to MoPubRewardedVideos.showRewardedVideo() results in this failure ["No video loaded for ad unit."](https://github.com/mopub/mopub-android-sdk/blob/6f8c457cf599fe2ae038abd6008b929ea5e48732/mopub-sdk/mopub-sdk-rewardedvideo/src/main/java/com/mopub/mobileads/MoPubRewardedVideoManager.java#L434)
### 3) Timing sensitive correct/incorrect response from MoPubRewardedVideos.hasRewardedVideo()
If MoPubRewardedVideos.hasRewardedVideo() is called in onRewardedVideoClosed() after an early exit and before MoPubRewardedVideos.loadRewardedVideo(), it returns true (incorrectly because MoPubRewardedVideos.showRewardedVideo() errors as explained in 2 above)
If waiting for a few seconds or calling MoPubRewardedVideos.loadRewardedVideo() first, MoPubRewardedVideos.hasRewardedVideo() returns false (correctly), making it difficult to know how to handle this scenario

### Expected outcome
When a user exits early, they should be able to display a Rewarded Ad again without receiving an error and there should be an additional related listener for onRewardedVideoClosedNotCompleted, which would help collect analytics/metrics

### Workaround
For 1 above, (not pretty and likely unreliable given it appears to be some type of timing condition), is to call MoPubRewardedVideos.loadRewardedVideo() inside of onRewardedVideoClosed() with a 2-3 second timing delay. Seems to work on an older Android API 22 device, as well as newer devices and emulators.
Example
```            
    Timer _t = new Timer();
    _t.schedule(new TimerTask() {
    @Override
    public void run() {
        runOnUiThread(new Runnable() {
            public void run() {                             
                MoPubRewardedVideos.loadRewardedVideo();
            }
        });
        }
    }, 2000);
```

#### Extra credit
Expected guidance: After a user views the first Rewarded Ad, when is the best time to call MoPubRewardedVideos.loadRewardedVideo() in the lifecycle to prepare for a second viewing?

Screenshots of the early exit dialog from AdMob - if the user selects "Skip" or "Close Video" then we have this issue, as explained above.
![image](https://user-images.githubusercontent.com/208641/48308862-cf55fa80-e521-11e8-86b8-11c0827c3d17.png)
![image](https://user-images.githubusercontent.com/208641/48309007-04b01780-e525-11e8-8f35-85b4fa7c2a2c.png)

**Comment by chauduyphanvu1** [comment]
@joeykrim, thanks for putting together the detailed report! To clarify, the behavior you are seeing should not be specific to AdMob rewarded video. We identified a couple SDK calls made that could be improved to provide a faster completion-to-request experience. It should resolve issues 1, 2, and 3 from your writeup. To confirm, making the next ad request in the `onRewardedVideoClosed()` callback is correct and should continued to be the best practice once we've released a fix for this issue.

Watch for the fix in the upcoming SDK release!

**Comment by nieny225** [comment]
👍👍👍

**Comment by chauduyphanvu1** [comment]
@joeykrim, we released SDK 5.4.1 yesterday that should have fixed this issue. I confirmed by daisy-chaining ad requests in the `onRewardedVideoClosed()` callback, and subsequent videos are able to load and play without a problem. Let us know if you continue seeing the issue!

Vu Chau
MoPub SDK Team
